pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
+	// Update updates the trie with a a prefix, key, and value.
+	//
+	// Note:If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Upsert(prefix uint, key K, value T)
+	// Delete removes a key with the given prefix and returns
+	// false if the key was not found.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Delete(prefix uint, key K) bool
+	// Lookup lookups any key that matches to at least
+	// the prefix argument.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Lookup(prefix uint, key K) T
+	// Len returns the number of entries in the Trie
+	Len() uint
+	// ForEach iterates over every element of the Trie.
+	ForEach(ForEachFunc[K, T])
+}
+
+// ForEachFunc is an interface for using the ForEach method of
+// all Trie types.
+type ForEachFunc[K, T any] func(prefix uint, key K, value T) bool
+
+// Key is an interface that implements all the necessary
+// methods to index and retrieve keys.
+type Key[K any] interface {
+	// CommonPrefix returns the amount of bits that
+	// are the same between this key and the argument
+	// value, starting from MSB.
+	CommonPrefix(K) uint
+	// BitValueAt returns the value of the bit at an argument
+	// index. MSB is 0 and LSB is n.
+	BitValueAt(uint) uint8
+	// Value returns the underlying value of the Key.
+	Value() K
+}
+
+// trie is the generic implementation of a bit-trie that can
+// accept arbitrary keys conforming to the Key[K] interface.
+type trie[K, T any] struct {
+	root      *node[K, T]
+	maxPrefix uint
+	entries   uint
+}
+
+// NewTrie returns a Trie that accepts the Key[K any] interface
+// as its key argument. This enables the user of this Trie to
+// define their own bit-key.
+func NewTrie[K any, T any](maxPrefix uint) Trie[Key[K], T] {
+	return &trie[K, T]{
+		maxPrefix: maxPrefix,
+	}
+}
+
+// node represents a specific key and prefix in the trie
+type node[K, T any] struct {
+	children     [2]*node[K, T]
+	prefixLen    uint
+	key          Key[K]
+	intermediate bool
+	value        T
+}
+
+// Lookup returns the longest or most "specific" key that matches the key
+// and prefix being looked up.
+//
+// Lookup starts traversal at the root key (or "node") in the trie.
+// The key and prefix being looked-up (the "lookup" key and prefix) are
+// compared to the a trie node's key and prefix (the "compare" key and
+// prefix) to determine the extent to which the keys match (from MSB to
+// LSB) up to the **least** specific (or lowest) prefix of the two keys
+// (for example, if one of the keys has a prefix of 2 and the other has
```

Comment:
by "prefix", do you mean "prefix length" here?

Reply:
Yes.